src/main/java/com/casper/sdk/model/clvalue/AbstractCLValue.java

```diff
@@ -65,17 +64,84 @@ protected void setJsonBytes(String bytes) {
+
+        return this.bytes;
+    }
+
     @JsonIgnore
     public abstract P getClType();
 
     public abstract void setClType(P value);
 
-    public abstract void serialize(SerializerBuffer ser, boolean encodeType) throws ValueSerializationException, NoSuchTypeException;
+    protected void serializePrefixWithLength(SerializerBuffer ser) throws ValueSerializationException {
+        SerializerBuffer localSer = new SerializerBuffer();
+        serialize(localSer);
+        int size = localSer.toByteArray().length;
+        ser.writeI32(size);
+    }
+
+    @Override
+    public AbstractCLValue<?, ?> deserialize(DeserializerBuffer deser, Target target) throws ValueDeserializationException, NoSuchTypeException {
+        if (target.equals(Target.BYTE)) {
+            return AbstractCLValue.createInstanceFromBytes(deser);
+        } else {
+            deserialize(deser);
+            return this;
+        }
+    }
+
+    public static AbstractCLValue<?, ?> createInstanceFromBytes(DeserializerBuffer deser) throws ValueDeserializationException, NoSuchTypeException {
+        int length = deser.readI32();
+        byte[] bytes = deser.readByteArray(length);
+        byte clType = deser.readU8();
+        try {
+            AbstractCLValue<?, ?> clValue = CLTypeData.getTypeBySerializationTag(clType).getClazz().getDeclaredConstructor().newInstance();
+            clValue.deserialize(new DeserializerBuffer(Hex.encode(bytes)));
+            return clValue;
+        } catch (InstantiationException | IllegalAccessException | NoSuchMethodException |
+                 InvocationTargetException e) {
+            throw new ValueDeserializationException("Error while instantiating CLValue", e);
+        }
+    }
+
+    @Override
+    public abstract void serialize(SerializerBuffer ser, Target target) throws ValueSerializationException, NoSuchTypeException;
 
+    @Override
     public abstract void deserialize(DeserializerBuffer deserializerBuffer) throws ValueDeserializationException;
 
-    public void encodeType(SerializerBuffer ser) throws NoSuchTypeException {
+    protected void encodeType(SerializerBuffer ser) throws NoSuchTypeException {
         byte val = (getClType().getClTypeData().getSerializationTag());
         ser.writeU8(val);
     }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (o == this) return true;
+        if (!(o instanceof AbstractCLValue)) return false;
+        final AbstractCLValue<?, ?> other = (AbstractCLValue<?, ?>) o;
+        if (!other.canEqual(this)) return false;
+        final Object this$bytes = this.getBytes();
+        final Object other$bytes = other.getBytes();
+        if (!Objects.equals(this$bytes, other$bytes)) return false;
+        final Object this$value = this.getValue();
+        final Object other$value = other.getValue();
+        if (this$value instanceof Map) {
+            return other$value.equals(other$value);
```

Comment:
This equals method is comparing with itself. Should be:
`return this$value.equals(other$value)`

Do we even need the ` if (this$value instanceof Map) ` could we not just call `return Objects.equals(this$value, other$value)` for all types?